text/0023-starlark_load_dataset.md

```diff
@@ -0,0 +1,249 @@
+```
+
+The default load dataset should be taken to mean "load this dataset with the _maximum available permission_". If a dataset is open, the user is free to read from any and all parts of the document.
+
+Combining a default of requesting maximum permission with _restrictive_ permission declared on the imported dataset supports our default stance of an open dataset ecosystem. The onus of declaring how a dataset can be used will fall to the owner of the dataset, and defaulting to max permissions forces the owner to assume that "if it's permitted, someone will use it", which is the correct assumption when attempting to restrict access.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+`load_dataset` has a fairly simple definition:
+
+```
+def load_dataset(reference): # return: dataset object
+```
+
+* **reference** is a string reference to a dataset, as defined by our dataset naming conventions elsewhere.
+* **dataset object** the returned value is a dataset object, as defined by ds.NewDataset in `qri-io/startf/ds`.
+
+As mentioned earler, the plan is to expand this definition with options once we've had time to do more research. Theses examples of valid reference strings, with `QmProfile` and `QmHash` being standins for full base-58 encoded hashes:
+
+```python
+load_dataset("b5/dataset")
+load_dataset("b5/dataset@QmProfile/ipfs/QmHash")
+load_dataset("@/ipfs/QmHash")
+```
+
+### `load_dataset` must be top level scope
+`load_dataset` statements will only be allowed within the primary scope of a transform script. Phrased another way: `load_dataset` _cannot_ be called within a function. This limits the scope of what is possible when composing dynamic import statements. This **won't** work:
+
+```python
+load("qri.star", "qri")
+
+# create a dataset of all dataset titles
+def transform(ds, ctx):
+  # error: load_dataset must be called in top level scope
+  datasets = [load_dataset(name) for name in qri.list_datasets()]
+  ds.set_body([ds.get_meta('title') for ds in datasets])
+```
+
+Statements that use scropts to compose import names, however, can and should work:
+
+```python
+pop_years = [load_dataset("b5/population_%s" % year) for year in range(2011,2017)]
+```
+
+The former example is making use of a _priveledged resource_ (the local qri node), while the latter is a convenience function for creating legal string names. For this assumption to hold, priveledged resources must _only_ be available within special functions, an assumption we already enforce.
+
+### Name Resolution
+When the starlark execution environment encounters a `load_dataset` statement, it has to _resolve_ the name, connecting the string reference to a dataset object. This may or may not require network access, depending on if the user currently has all of the specified versions locally. All naming duties required by `load_dataset` will be offloaded to Qri's name resolution system, but a few interactions merit calling out here:
```

Comment:
Specify that if a dataset cannot be loaded, or doesn't exist anywhere, bail out and exit the script. There shouldn't be a recovery mechanism, which would allow probing datasets to see whether they exist or not.